ethereum/contracts/IncentivizedReceiveChannel.sol

```diff
@@ -0,0 +1,139 @@
+        uint256 parachainBlockNumber,
+        bytes memory ourParachainMerkleProof,
+        bytes memory parachainHeadsMMRProof
+    ) public {
+        verifyCommitment(
+            commitment,
+            commitmentContents,
+            ourParachainMerkleProof,
+            parachainHeadsMMRProof
+        );
+        processCommitmentContents(commitmentContents);
+    }
+
+    function verifyCommitment(
+        Commitment memory commitment,
+        CommitmentContents memory commitmentContents,
+        bytes memory ourParachainMerkleProof,
+        bytes memory parachainHeadsMMRProof
+    ) internal returns (bool success) {
+        // Prove we can get the MMRLeaf that is claimed to contain our Parachain Block Header
+        // BEEFYLightClient.verifyMMRLeaf(parachainHeadsMMRProof)
+        // BeefyLightClient{
+        // verifyMMRLeaf(parachainHeadsMMRProof) {
+        //MMRVerification.verifyInclusionProof(latestMMRRoot, parachainHeadsMMRProof)
+        // }
+        //}
+        //}
+        // returns mmrLeaf;
+
+        // Prove we can get the claimed parachain block header from the MMRLeaf
+        // allParachainHeadsMerkleTreeRoot = mmrLeaf.parachain_heads;
+        // MerkeTree.verify(allParachainHeadsMerkleTreeRoot, ourParachainMerkleProof)
+        // returns parachainBlockHeader
+
+        // Prove that the commitment is in fact in the parachain block header
+        // require(parachainBlockHeader.commitment == commitment)
+
+        // Prove that the commitmentContents match the commitment
+        // require(commitment == hash(commitmentContents))
+
+        // Require there is enough gas to play all messages
+        require(
+            gasleft() >=
+                commitmentContents.messages.length * MAX_GAS_PER_MESSAGE,
+            "insufficient gas for delivery of all messages"
+        );
+
+        // Require all payloads are smaller than max_payload_size
+        for (uint256 i = 0; i < commitmentContents.messages.length; i++) {
+            require(
+                commitmentContents.messages[i].payload.length <=
+                    MAX_PAYLOAD_BYTE_SIZE,
+                "message payload bytesize exceeds maximum payload size"
+            );
+        }
+        return true;
+    }
+
+    function processCommitmentContents(
+        CommitmentContents memory commitmentContents
+    ) internal {
+        for (uint256 i = 0; i < commitmentContents.messages.length; i++) {
+            // Check message nonce is correct and increment nonce for replay protection
+            Message memory message = commitmentContents.messages[i];
+            require(message.nonce == lastProcessedNonce + 1, "invalid nonce");
+
+            lastProcessedNonce = lastProcessedNonce + 1;
+
+            // Deliver the message to the destination
+            // Delivery will have fixed maximum gas allowed for the destination app.
+            bytes memory callInput = message.payload;
```

Comment:
Following on from our convo this morning - I suspect you are wanting to implement a generic RPC protocol by including the method ID in the message payload, in essence allowing a substrate application pallet to call any method on an Ethereum contract.

I think some kind of generic RPC like this can be useful but there are a few considerations we need to take into account. 

Keen to hop onto a call later when you're ready to discuss this.

Reply:
I'd like to join the call as well, the considerations are probably relevant to ongoing work on https://github.com/Snowfork/polkadot-ethereum/pull/228

Reply:
Can reference https://github.com/Snowfork/polkadot-ethereum/pull/228 for the proper payload encoding. The payload byte contents are:
- the function signature. This is the first 8 bytes of the result of `keccak256(functionName(argVariableType1,argVariableType2,...))`
- each variable padded to 32 bytes

An example ETHApp Unlock function payload is `0xed581fd300000000000000000000000022448f19a3cb4ee570c2ec6d3323761d3399bbbd0000000000000000000000000000000000000000000000000000000000000002`
- Function signature: `0xed581fd3` derived from `keccak256(Unlock(address, uint256))`
- Recipient address: `00000000000000000000000022448f19a3cb4ee570c2ec6d3323761d3399bbbd`
- Amount: `0000000000000000000000000000000000000000000000000000000000000002`

Reply:
As I've discussed with Denali on Slack, the ability to call any method on a target contract does change the design a bit, and that leads to inconsistencies in several areas:

* On the Substrate side, apps still need to implement a `handle` method. Theoretically yes we can update the design to support calling any dispatchable, but practically that's not going to be implemented any time soon. 
* One suggested use-case for the proposed feature was to have the bridge support Ethereum contracts which are already deployed and live, and of course don't implement `handle` methods. I can see where the feature helps with that. However we need to consider the opposite direction where these old previously-deployed contracts need to submit mesages to the Send Channel contract. That's currently not possible without deploying an intermediary adaptor contract. But now we have this inconsistency where the adaptor isn't needed in the inbound direction, and only in the outbound direction. This adaptor contract needs to somehow fit in with the design and I don't think we have thought that out enough.

Then apart from the above design inconsistencies, there are other concerns:
* Substrate applications now need to know what method to call on their peer contract, so this adds complexity and configuration to the Substrate side where previously none existed.
* Previously, end-users who want to develop Ethereum apps on our bridge would have only needed to write a contract that implements an `IApplication` interface with a `handle` function. This makes things simpler for users as they know what they need to implement. They just need to fill in an implementation for `handle` and its done.

Given all the above, I suggest we limit the scope of this PR (incentivized channel) and defer this message dispatch feature until later on when we've reached some kind of consensus on the above.

Reply:
> i think these inconsistencies are totally fine, don't see any major issue with them.

Its hard to reply to that. Design can be subjective.

Shouldn't this PR be focused on receive channel stuff? The proposed message dispatch feature is really orthogonal to that I think.

> Not really - our pallets already have configuration/complexity requirements, i don't think this adds much more?

It still adds up though.

> Huh? No, now they are not constrained to any interface, they can develop ethereum apps with solidity as normal - seems much simpler?

Nope, I still think it makes it harder for users for the reasons I mentioned in my original point.